src/main/java/dev/blaauwendraad/masker/json/JsonPathSegmentReference.java

```diff
@@ -0,0 +1,48 @@
+package dev.blaauwendraad.masker.json;
+
+/**
+ * A mutable reference to a sequence of bytes in <code>dev.blaauwendraad.masker.json.MaskingState#message</code>. It is used to represent json path segments.
+ * <p>
+ * A reference is represented by a pair of integers:
+ * <ul>
+ *  <li><code>offset</code> denotes the start index of a segment in the message. In case the segment is an array, <code>offset</code> denotes the element index.</li>
+ *  <li><code>length</code> denotes the length of a segment in the message. In case the segment is an array, <code>length</code> is set to -1.</li>
+ * </ul>
+ */
+public class JsonPathSegmentReference {
```

Comment:
Maybe that's a good candidate for sealed interface: `ArrayPathReference(int index)` and `NodePathReference(offset, length)`?

Reply:
@donavdey I've pushed a commit making it sealed, let me know if it looks worse to you and we can revert it

Reply:
This looks good. Now I finally see how sealed interfaces are useful in the first place.